File: pipeline/src/main/java/dev/mzarnowski/system/pipeline/ForEach.java
package dev.mzarnowski.system.pipeline;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

final class ForEach<A> extends Sink {
    private final Pipeline owner;
    private final Reader<A> reader;
    private final Consumer<A> consumer;

    private final AtomicBoolean isDisposed = new AtomicBoolean(false);

    public ForEach(Pipeline owner, Reader<A> reader, Consumer<A> consumer) {
        super(owner.scheduler);
        this.owner = owner;
        this.reader = reader;
        this.consumer = consumer;
    }

    @Override
    public void dispose() {
        if (isDisposed.compareAndSet(false, true)) {
            super.dispose();
        }
    }

    @NotNull
    protected Result iterate() {
        // TODO create observer class, which is not requesting eagerly and is disposed automatically
        var requested = reader.request();

        var available = reader.claim(1, owner.batchSize);
        for (int i = 0; i < available; i++) {
            var next = reader.get(i);
            consumer.accept(next);
        }

        reader.release(available);
        if (requested) return Wait.INSTANCE;
        if (available == 0) return Break.INSTANCE;
        return Continue.INSTANCE;
    }
}

File: pipeline/src/main/kotlin/dev/mzarnowski/system/pipeline/ForEach.kt
package dev.mzarnowski.system.pipeline

import java.util.concurrent.ScheduledExecutorService as Scheduler

class ForEach<A>(scheduler: Scheduler, val reader: Reader<A>, val f: (A) -> Unit) : Task(scheduler) {
    override fun iterate(): Result {
        val available = reader.claim(1, 32)
        if (available < 0) return Break

        reader.request() // eagerly request
        if (available > 0) {
            repeat(available) {
                val next = reader.read(it)
                f(next)
            }

            reader.release(available)
        }

        return Continue
    }
}